text/0000-floating-point-additional-precision.md

```diff
@@ -0,0 +1,141 @@
+behavior from within code, such as `#[disable_extra_fp_precision]`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+If Rust already provided bit-for-bit identical floating-point computations
+across platforms, this change could potentially allow floating-point
+computations to differ by platform (though never below the standards-required
+accuracy). However, standards-compliant implementations of math functions on
+floating-point values may already vary slightly by platform, sufficiently so to
+produce different binary results. This proposal can never make results *less*
+accurate, it can only make results *more* accurate.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+We could provide a separate set of types and allow extra accuracy in their
+operations; however, this would create ABI differences between floating-point
+functions, and the longer, less-well-known types seem unlikely to see
+widespread use.
+
+We could provide an option to enable extra accuracy for the default
+floating-point types, but disable it by default. This would leave the majority
+of floating-point code unable to use these optimizations, however; defaults
+matter, and the majority of code seems likely to use the defaults.
+
+We could do nothing, and require code to use `a.mul_add(b, c)` for
+optimization; however, this would not allow for similar future optimizations,
+and would not allow code to easily enable this optimization without substantial
+code changes.
+
+We could narrow the scope of optimization opportunities to *only* include
+floating-point contraction but not any other precision-increasing operations.
+See the [future possibilities](#future-possibilities) section for further
+discussion on this point.
+
+# Prior art
+[prior-art]: #prior-art
+
+This has precedent in several other languages and compilers:
+
+- [C11](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf) allows
+  this with the `STDC FP_CONTRACT` pragma enabled, and the default state
+  of that pragma is implementation-defined. GCC enables this pragma by
+  default, [as does the Microsoft C
+  compiler](https://docs.microsoft.com/en-us/cpp/preprocessor/fp-contract?view=vs-2019).
+
+- [The C++ standard](http://eel.is/c++draft/expr.pre#6) states that "The
+  values of the floating operands and the results of floating
+  expressions may be represented in greater precision and range than
+  that required by the type; the types are not changed thereby."
+
+- The [Fortran standard](https://www.fortran.com/F77_std/rjcnf0001-sh-6.html#sh-6.6.4)
+  states that "the processor may evaluate any mathematically equivalent
+  expression", where "Two arithmetic expressions are mathematically
+  equivalent if, for all possible values of their primaries, their
+  mathematical values are equal. However, mathematically equivalent
+  arithmetic expressions may produce different computational results."
```

Comment:
I'm not familiar with Fortran (or at least this aspect of it), but this quote seems to license far more than contraction, e.g. all sorts of `-ffast-math` style transformation that ignore the existence of NaNs. Is that right?

Reply:
@rkruppe That's correct, Fortran also allows things like reassociation and commutation, as long as you never ignore parentheses.